codalab/worker/worker.py

```diff
@@ -145,8 +145,9 @@ def __init__(
             bundle_runtime=bundle_runtime,
         )
 
-        # Lock ensures .checkin() is not run concurrently.
-        self._checkin_lock = Lock()
+        # Lock ensures listening thread and main thread don't simultaneously
+        # access the runs dictionary, thereby causing race conditions.
+        self._lock = RLock()
```

Comment:
Why use RLock instead of Lock?

Reply:
It's just for elegance's sake.

RLock allows the lock to be acquired multiple times (provided it is also released precisely as many times at is acquired). This is desirable in this case because, to avoid race conditions, we must call self.process_runs _before_ the lock is released within the checkin() function. By using RLock, we allow the code to acquire the lock a second time so that we can run process_runs within checkin.

If we didn't do this, we could use a flag to tell process_runs we already acquired the lock and not acquire the lock in that case, but that seemed less elegant to me.